File: app/src/main/java/com/stackanswer/tidakdigunakan/DetailMovieViewModelFactory.java
package com.stackanswer.tidakdigunakan;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class DetailMovieViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private static volatile DetailMovieViewModelFactory INSTANCE;

    private final DetailMovieRepository mDetailMovieRepository;

    private DetailMovieViewModelFactory(DetailMovieRepository academyRepository) {
        mDetailMovieRepository = academyRepository;
    }

    public static DetailMovieViewModelFactory getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (DetailMovieViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new DetailMovieViewModelFactory(Injection.provideDetailMovieRepository(context));
                }
            }
        }
        return INSTANCE;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass.isAssignableFrom(DetailMovieViewModel.class)) {
            return (T) new DetailMovieViewModel(mDetailMovieRepository);
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: app/src/main/java/com/stackanswer/viewmodel/factory/DetailMovieViewModelFactory.kt
//package com.stackanswer.viewmodel.factory
//
//import android.content.Context
//import androidx.lifecycle.ViewModel
//import androidx.lifecycle.ViewModelProvider
//import com.stackanswer.core.domain.usecase.DetailMovieUseCaseKt
//import com.stackanswer.injection.Injection
//import com.stackanswer.viewmodel.DetailMovieViewModelKt
//
//class DetailMovieViewModelFactory private constructor(private val tourismUseCase: DetailMovieUseCaseKt) :
//    ViewModelProvider.NewInstanceFactory() {
//
//    companion object {
//        @Volatile
//        private var instance: DetailMovieViewModelFactory? = null
//
//        fun getInstance(context: Context): DetailMovieViewModelFactory =
//            instance ?: synchronized(this) {
//                instance ?: DetailMovieViewModelFactory(
//                    Injection.provideDetailMovieUseCase(context)
//                )
//            }
//    }
//
//    @Suppress("UNCHECKED_CAST")
//    override fun <T : ViewModel> create(modelClass: Class<T>): T =
//        when {
//            modelClass.isAssignableFrom(DetailMovieViewModelKt::class.java) -> {
//                DetailMovieViewModelKt(tourismUseCase) as T
//            }
//            else -> throw Throwable("Unknown ViewModel class: " + modelClass.name)
//        }
//}